source/mainnet/smart-contracts/guides/unit-test-contract.rst

```diff
@@ -339,18 +339,144 @@ state and balance fields:
        );
        ...
 
-.. warning::
+Reentrancy
+----------
+
+You should watch out for *reentrancy problems*, which can occur when calls to
+``invoke_contract`` end up updating the state of your own contract.
+
+.. code-block:: rust
+
+    let state_copy = *host.state();
+    host.invoke_contract(...);
+
+    // *host.state() and state_copy might not be equal any more due to reentrancy.
+    do_something_with(state_copy);
+
+Another example of reentracny is when the state is *not* updated properly before making an external call.
+This can lead to reentrant calls that pass some validation that is based on the current state, while these calls should fail.
```

Comment:
```suggestion
This can lead to reentrant calls that pass some validation that is based on the current state, even though these calls should fail.
```